File: Kristabel Wong/wk01 - starts 7th Mar/5-fri/MTA Lab/js/MTA-Lab.js
// # MTA Lab

// ## Objectives:
// * Apply your knowledge of Javascript to solve a real world problem.
// * Get really good at array manipulation.

// #### Activity
// * Create a program that models a simple subway system.

// * The program takes the line and stop that a user is getting on at and the line
// and stop that user is getting off at and prints the journey and the total number of stops for the trip in the console:

// ```javascript
// planTrip('N', 'Times Square', '6', '33rd'); // This is only a suggested function name and signature.

// // console.log() shows output similar to this:
// // "You must travel through the following stops on the N line: 34th, 28th, 23rd, Union Square."
// // "Change at Union Square."
// // "Your journey continues through the following stops: 23rd, 28th, 33rd."
// // "7 stops in total."
// ```

// * There are 3 subway lines:
//   * The N line has the following stops: Times Square, 34th, 28th, 23rd, Union Square, and 8th
//   * The L line has the following stops: 8th, 6th, Union Square, 3rd, and 1st
//   * The 6 line has the following stops: Grand Central, 33rd, 28th, 23rd, Union Square, and Astor Place.
//   * All 3 subway lines intersect at Union Square, but there are no other intersection points. (For example, this means the 28th stop on the N line is different than the 28th street stop on the 6 line, so you'll have to differentiate this when you name your stops in the arrays.)
// * Tell the user the number of stops AND the stops IN ORDER that they will pass through or change at.


// #### Hints:
// * Work out how you would do it on paper first! Then start to explain that process in Javascript.
// * Get the program to work for a single line before trying to tackle multiple lines.
// * Don't worry about prompting the user for input. Hard code some values to get it working. You can use ```prompt()``` later to make it more interactive.
// * Consider diagramming the lines by sketching out the subway lines and their stops and intersection.
// * The key to the lab is finding the index positions of each stop. (hint: ```indexOf()```)
// * Make sure the stops that are the same for different lines have different names (i.e. 23rd on the N and on the 6 need to be differentiated)

// let N = ['Times Square', '34th', '28th', '23rd', 'Union Square', '8th'];
// let L = ['8th', '6th', 'Union Square', '3rd', '1st'];
// let six = ['Grand Central', '33rd', '28th', '23rd', 'Union Square', 'Astor Place'];


const trainLines = {

    lines: { 
        'N': ['Times Square', '34th', '28th', '23rd', 'Union Square', '8th'],
        'L': ['8th', '6th', 'Union Square', '3rd', '1st'],
        '6': ['Grand Central', '33rd', '28th', '23rd', 'Union Square', 'Astor Place']
    }

};

// console.log(`The stops are ${trainLines.lines['N'].join(', ')}`);
// console.log(trainLines.lines['L']);
// console.log(trainLines.lines['6']);
// console.log(" ");
// console.log(trainLines.lines['L'].indexOf('none')); // --> this gives me a negative index 
// console.log(trainLines.lines['N'][2]);
// console.log(trainLines.lines['6'][5]);


const planTrip = function (line, firstStop, line2, endStop ) {
    let train = trainLines.lines[line];
    let train2 = trainLines.lines[line2];
    let intersect = 'Union Square'
    let totalStops = 0;

    if (line !== 'N' && line !== 'L' && line !== '6') { // if the train line does not exist  --> how do i make this dynamic?
        console.log(`The train line ${line} does not exist, please enter in train line N, L or 6.`);
        
    } else if (line2 !== 'N' && line2 !== 'L' && line2 !== '6') { // if the train line does not exist
        console.log(`The train line ${line2} does not exist, please enter in train line N, L or 6.`);

    } else if (train.indexOf(firstStop) === -1) {   // if the train stop does not exist on the first line
        console.log(`The beginning stop ${firstStop} does not exist on line ${line}. Please select from: ${train.join(', ')}.`);

    } else if (train2.indexOf(endStop) === -1) { // if the train stop does not exist on the second line
        console.log(`The destination stop ${endStop} does not exist on line ${line2}. Please select from: ${train2.join(', ')}.`);

    } else if (line === line2) {
        let stop = train.indexOf(firstStop);
        let stop2 = train.indexOf(endStop);
        let rideTo = [];

        if (stop < stop2) {  // this is for when you are going down the line
            for (i = stop+1; i < stop2+1; i++) {  
                rideTo.push(train[i]);
                totalStops += 1;
            }
            console.log(`You must travel through the following stops on the ${line} line: ${rideTo.join(', ')}.`);
            console.log(`${totalStops} stops in total.`)

        } else if (stop > stop2) {
            for (i = stop-1; i > stop2-1; i--) {  // this is for when you are going back up the line
                rideTo.push(train[i]);
                totalStops += 1;
            }
            console.log(`You must travel through the following stops on the ${line} line: ${rideTo.join(', ')}.`);
            console.log(`${totalStops} stops in total.`)

        } else if (stop === stop2) {  // if the user has enter the same beginning and ending stop on the same line
            console.log(`You have already arrived at your destination - stop ${endStop} on line ${line2}.`);
        } 

    } else if (line !== line2) {
        let stop = train.indexOf(firstStop);
        let stop2 = train.indexOf('Union Square');
        let rideTo = [];
        
        for (i = stop+1; i < stop2+1; i++) { // this is to the intersecting station (Union Square)
            rideTo.push(train[i]);
            totalStops += 1;
        }
        console.log(`You must travel through the following stops on the ${line} line: ${rideTo.join(', ')}.`)
        console.log(`Change at ${intersect}.`);
        
        let stop3 = train2.indexOf('Union Square'); // this is from the intersecting station (Union Square)
        let stop4 = train2.indexOf(endStop);
        let remainingStops = [];

        if (stop3 < stop4) {  // this is for when you are going down the line
            for (i = stop3+1; i < stop4+1; i++) {
                remainingStops.push(train2[i]);
                totalStops += 1;
            }
            console.log(`Your journey continues through the following stop: ${remainingStops.join(', ')}.`);
            console.log(`${totalStops} stops in total.`)

        } else {
            for (i = stop3-1; i > stop4-1; i--) {  // this is for when you are going back up the line
                remainingStops.push(train2[i]);
                totalStops += 1;
            }
            console.log(`Your journey continues through the following stop: ${remainingStops.join(', ')}.`);
            console.log(`${totalStops} stops in total.`)
        }

    }

};
        

planTrip('N','8th','N','Times Square');
console.log(' '); // --> this makes it easier to read in the console
planTrip('N', 'Times Square', '6', '33rd');
console.log(' '); // --> this makes it easier to read in the console
planTrip('L', '6th', '6', 'Grand Central');
console.log(' '); // --> this makes it easier to read in the console
planTrip('N','28th','6','28th');
console.log(' '); // --> this makes it easier to read in the console
console.log('ERROR TESTING'); // --> this makes it easier to read in the console
planTrip('L', '6th', 'L', '6th'); // --> same beginning and ending stop
console.log(' '); // --> this makes it easier to read in the console
planTrip('N', 'Times Square', '6', 'Tim Squa'); // to test errors
console.log(' '); // --> this makes it easier to read in the console
planTrip('N', 'Times Square', '8', 'Tim Squa'); // to test errors
// console.log(Object.keys(trainLines.lines).join(', ')); --> key names
// console.log(Object.keys(trainLines.lines).length); --> shows 3. how do i use this to make a dynamic formula?

File: Joseph Hu/wk04 - starts 28th Mar/2-tue/MTAlab.rb
#Hash containing lines as keys and arrays of stops as values
SubwayMap = {
    "N"=> ["Times Square", "34th", "28th", "23rd", "Union Square", "8th"],
    "L" => ["8th", "6th", "Union Square", "3rd", "1st"],
    "6" => ["Grand Central", "33rd", "28th", "23rd", "Union Square", "Astor Place"]
};

def planTrip 
    #displays subway map
    puts "Here is the subway Map:"

    SubwayMap.each {|key, value| puts "stations on Line #{key} are: #{value}"}

    # asks for user input for lines and stations
    puts "Which line are you on?"
    current_line = gets.upcase.chomp
    puts "Which station are you travelling from?"
    current_station = gets.chomp
    puts "Which line would you like to travel to?"
    dest_line = gets.upcase.chomp
    puts "Which station would you like to travel to?"
    dest_station = gets.chomp

    #ensures that user enters a valid line
    if SubwayMap.has_key?(dest_line) && SubwayMap.has_key?(current_line)

        #ensures that the user enters a valid station
        if SubwayMap[current_line].include?(current_station) && SubwayMap[dest_line].include?(dest_station) 

            #my destination message and continue message
            dest_Message = "You must travel through the following stops on the #{current_line} line: "
            continue_Message = "Your journey continues through the following stops on the #{dest_line} line: "

            #my arrays to append my stations from my first line and stations from my second line 
            line1_stops =[]
            line2_stops =[]

            #Array (values) of my current and destination lines
            line1 = SubwayMap[current_line]
            line2 = SubwayMap[dest_line]

            #Indexes of Union Square for both lines
            union_square_index_1 = line1.index("Union Square")
            union_square_index_2 = line2.index("Union Square")

            #Index of starting and destination stations
            d1Index = line1.index(current_station)
            d2Index = line2.index(dest_station)

            #if my lines are not the same line
            if current_line != dest_line

                #if my starting station is before union Square, iterate forward until i reach Union Square
                if d1Index < union_square_index_1
                    line1[d1Index+1..union_square_index_1].each{|station| line1_stops << station}
                end

                #if my destination station is before destination Union Square, iterate in reverse until the destination from Union Square.
                if d2Index < union_square_index_2
                    line2[d2Index..union_square_index_2-1].reverse_each{|station| line2_stops << station }
                end

                #if my starting station is after union Square, iterate backwards until i reach union square
                if d1Index > union_square_index_1
                    line1[union_square_index_1..d1Index-1].reverse_each{|station| line1_stops << station}
                end

                #if my destination station is after my destination union square index, iterate forwards until i reach the station.
                if d2Index > union_square_index_2
                    line2[union_square_index_2+1..d2Index].each{|station| line2_stops << station}
                end

                # take all the stations pushed from my line 1 iterations and combine it with the destination message.
                dest_Message = dest_Message + line1_stops.join(", ") + "."

                # take all the stations pushed from my line 2 iterations and combine it with the continue message.
                continue_Message = continue_Message + line2_stops.join(", ") + "."

                #if my starting station is union square, i can change at Union Square right away and just put the continue message.
                if d1Index == union_square_index_1
                    puts "Change at Union Square."
                    puts continue_Message

                #if my destination station is union square then i just need to travel up to union square and change
                elsif d2Index == union_square_index_2
                    puts dest_Message
                    puts "Change at Union Square."

                #else, my lines are different and stations are different so put the dest and cont messages.
                else
                    puts dest_Message
                    puts "Change at Union Square."
                    puts continue_Message
                end

            #if i'm travelling on the same line  
            else
                #if my starting station is before my destination iterate until i reach my destination
                if d1Index < d2Index
                    line1[d1Index+1..d2Index].each{|station| line1_stops << station }
                #if my destination station before my starting station iterate backwards until i reach my destination.
                else
                    line1[d2Index..d1Index-1].reverse_each{|station| line1_stops << station}
                end
                #put out my destination message by combining the line1 array with my mesage.
                dest_Message = dest_Message + line1_stops.join(", ") + "."
                puts dest_Message
            end

            #if my lines and start/destination are the same, then say i'm already here
            if (current_line == dest_line && current_station == dest_station)
                puts "You are already here!"
            end

            # outputs the total number of stops
            puts "#{line1_stops.length + line2_stops.length} stops in total."
            
            #reprompts for a trip
            puts "Would you like to plan another trip? Y/N"
            restart = gets.chomp
            if restart == "Y"
                planTrip
            end

        #if my stations don't exist reprompt
        else
            puts "Please enter a valid station!"
            planTrip
        end

    #if my lines don't exist reprompt
    else
        puts "Please enter a valid line!"
        planTrip    
    end
end

# call the function 
planTrip